packages/lib/src/interactions/SelectionTool.tsx

```diff
@@ -46,69 +62,81 @@ function SelectionTool(props: Props) {
 
       const { target, pointerId } = sourceEvent;
       (target as Element).setPointerCapture(pointerId);
-      setStartPoint(worldToData(unprojectedPoint));
+      setSelection({
+        startPoint: worldToData(unprojectedPoint),
+        endPoint: worldToData(unprojectedPoint),
+      });
+      setOriginPoint(new Vector2(unprojectedPoint.x, unprojectedPoint.y));
 
       if (onSelectionStart) {
         onSelectionStart();
       }
     },
-    [onSelectionStart, shouldInteract, worldToData]
+    [onSelectionStart, setSelection, shouldInteract, worldToData]
   );
 
   const onPointerMove = useCallback(
     (evt: CanvasEvent<PointerEvent>) => {
-      if (!startPoint) {
+      if (!originPoint) {
         return;
       }
 
       const { unprojectedPoint, sourceEvent } = evt;
-      const point = worldToData(boundPointToFOV(unprojectedPoint, camera));
-      setEndPoint(point);
+      const worldEndPoint = new Vector2(unprojectedPoint.x, unprojectedPoint.y);
+      const worldStartPoint = getWorldStartPoint(originPoint, worldEndPoint);
+      const selection = {
+        startPoint: worldToData(boundPointToFOV(worldStartPoint, camera)),
+        endPoint: worldToData(boundPointToFOV(worldEndPoint, camera)),
+      };
+      setSelection(selection);
 
       if (onSelectionChange && shouldInteract(sourceEvent)) {
-        onSelectionChange({
-          startPoint,
-          endPoint: point,
-        });
+        onSelectionChange(selection);
       }
     },
     [
-      startPoint,
+      originPoint,
+      getWorldStartPoint,
       worldToData,
       camera,
-      setEndPoint,
+      setSelection,
       onSelectionChange,
       shouldInteract,
     ]
   );
 
   const onPointerUp = useCallback(
     (evt: CanvasEvent<PointerEvent>) => {
-      if (!startPoint) {
+      if (!originPoint) {
         return;
       }
 
       const { sourceEvent, unprojectedPoint } = evt;
       const { target, pointerId } = sourceEvent;
       (target as Element).releasePointerCapture(pointerId);
 
-      setStartPoint(undefined);
-      setEndPoint(undefined);
+      const worldEndPoint = new Vector2(unprojectedPoint.x, unprojectedPoint.y);
+      const worldStartPoint = getWorldStartPoint(originPoint, worldEndPoint);
+      const selection = {
+        startPoint: worldToData(boundPointToFOV(worldStartPoint, camera)),
+        endPoint: worldToData(boundPointToFOV(worldEndPoint, camera)),
+      };
```

Comment:
This part is the same as the one in `onPointerMove`. It might be a way to reduce the complexity of the component (which starts to be quite bloated).